### ISSUE TITLE
How does secret sharing work with one source

### ISSUE THREAD

**Issue by meryemJanatiIdrissi** [facebookresearch/CrypTen#188]
Hello
in the following example, x is encrypted though we have only one source. How is that possible, don't we need at least two sources? else how does it work?

```
x = torch.tensor([1.0, 2.0, 3.0])

x_enc = crypten.cryptensor(x)
```
thank you in advance

**Comment by gmuraru** [comment]
Hello!
This is an answer based on my understanding of how CrypTen works! - someone from the CrypTen team can correct me if I am wrong :smile:

**More than 1 Party**
In the usual scenario, you should have more than 1 party in a SMPC.
CrypTen is using *PRZS (Pseudo Random Zero share)* when sharing a value.
For doing this, each "party" holds 2 generators and one of them is "shared" with the previous party.

![party_1](https://user-images.githubusercontent.com/7805588/98097828-364ee580-1e96-11eb-9d96-bc40df615e16.png)

Then the parties generate shares (one of the shares is not "final") using those two generators.
![party_2](https://user-images.githubusercontent.com/7805588/98098760-6ea2f380-1e97-11eb-8117-54357f5f5284.png)

Where:
![exp](https://user-images.githubusercontent.com/7805588/98098776-72367a80-1e97-11eb-88e6-e8febb192a63.png)
As you can see the shares would have the sum 0 (hence the name Pseudo Random Zero Share)

Now, the party that has the **secret** adds its personal share. Now, when combining the shares (the reconstruction of the secret value) - you will get the **secret**

**1 Party**
In the case there is only one party:
![sg_share](https://user-images.githubusercontent.com/7805588/98099237-f983ee00-1e97-11eb-82ba-bced7afcbe23.png)
The constructed share would be simply: ```0 + secret```.
You need to take into consideration that the secret, is "transformed" using a *FixedPointEncoder*.

CrypTen uses as default precision 16 bits --> which gives us a constant of ```2^16``` that would play the role of a *scale*.

** Example**
Let's say you have:
```
x = torch.tensor([1.0, 2.0, 3.0])
x_enc = crypten.cryptensor(x)
```
With the generators, we created a "zero value" that we add with the fixed-point representation of our tensor.
It would simply result in a single share that is exactly our fixed-point number: ```x_enc = [1.0 * 2^16, 2.0 * 2^16, 3.0 * 2^16]``` (where 2^16 is the scale).

So you should get back:
[65536, 131072, 196608]

**TL;DR**: You get back only the fixed-point representation of that tensor.

**Comment by gmuraru** [comment]
@meryemJanatiIdrissi did the post above answer your question?

**Comment by lvdmaaten** [comment]
@gmuraru The description is completely accurate 👍 

In the single-party setting, there is essentially no encryption happening. This makes sense because if only one party is going to participate in the computation, why bother encrypting data? (It is called secure *multi*-party computation for a reason :).)

In practice, you will always run CrypTen with two or more parties. Note that in that case, running `x_enc = crypten.cryptensor(x)` is actually running `x_enc = crypten.cryptensor(x, src=0)`. This means that the party with rank 0                   is providing the `x` tensor to be encrypted; it is the party that will add `x` to its part of the PRZS. You can change the value of `src` if you want another party to provide the `x` for encryption.

**Comment by n0pn0pn0p** [comment]
In the two-parties setting，P1 and P2 both know the G1 and G2,  if  it's not security since P1 and P2 can get the secret?

**Comment by lvdmaaten** [comment]
In the two-party case, party 1 will have generator 1 and 2 whereas party 3 will have generator 2 and 3.

**Comment by louislouislouislouis** [comment]
> Hello! This is an answer based on my understanding of how CrypTen works! - someone from the CrypTen team can correct me if I am wrong 😄
> 
> **More than 1 Party** In the usual scenario, you should have more than 1 party in a SMPC. CrypTen is using _PRZS (Pseudo Random Zero Share)_ when sharing a value. For doing this, each "party" holds 2 generators and one of them is "shared" with the previous party.
> 
> ![party_1](https://user-images.githubusercontent.com/7805588/98097828-364ee580-1e96-11eb-9d96-bc40df615e16.png)
> 
> Then the parties generate shares (one of the shares is not, hmm...let's say "final", you will see down below why I say that) using those two generators. ![party_2](https://user-images.githubusercontent.com/7805588/98098760-6ea2f380-1e97-11eb-8117-54357f5f5284.png)
> 
> Where: ![exp](https://user-images.githubusercontent.com/7805588/98101866-795f8780-1e9b-11eb-9006-89ead3839d92.png) As you can see the shares would have the sum 0 (hence the name Pseudo-Random Zero Share) Now, the party that shares the **secret**, adds this value to its personal share. When you combine the shares (the reconstruction of the initial data) - you will get back the **secret**.
> 
> **1 Party** In the case there is only one party: ![sg_share](https://user-images.githubusercontent.com/7805588/98101911-8c725780-1e9b-11eb-8373-c3e7e0ae178f.png)
> 
> The constructed share would be simply: `0 + secret`. ^^ The zero value is because you have the same "generator" (more exactly you initialized the 2 generators using the same seed). The "next party", from which we should have gotten one of the seeds is "party 1" (so we get the exact same seed ==> the generators would generate the same series of random numbers) You need to take into consideration that the secret, is "transformed" using a _FixedPointEncoder_.
> 
> CrypTen uses as default precision 16 bits --> which gives us a constant of `2^16` that would play the role of a _scale_.
> 
> **Example** Let's say you have:
> 
> ```
> x = torch.tensor([1.0, 2.0, 3.0])
> x_enc = crypten.cryptensor(x)
> ```
> 
> With the generators, we created a "zero value" that we add with the fixed-point representation of our tensor. It would simply result in a single share that is exactly our fixed-point number: `x_enc = [1.0 * 2^16, 2.0 * 2^16, 3.0 * 2^16]` (where `2^16` is the scale).
> 
> So you should get back: `[65536, 131072, 196608]`
> 
> **TL;DR**: You get back only the fixed-point representation of that tensor.

When generating Share1, Share2 and Share3, what is the function next, involved in next(G1),next(G2) and next(G3). Is the next function a this function a PRF, like HMAC?
